Add new tag !include_dir_merge_recursive

Phist0ne opened the pull request:
## Proposed change
Enable the user to split up the configuration files in a way, that allows for the complete context to be present in every file. Also enable the user to group all configuration for a single device in one file even if it contains configuration for multiple domains, e.g. switches, sensors and home-assistant.customize customization.
At the moment with the existing include tags the configuration for different domains (sensor, switch, component specific) needs to be split between different folders / files.
One may attempt something similar using the existing !include_dir_merge_named tag, by using one folder per device containing files like sensor.yaml and switch.yaml, but the last file for every domain that is loaded will shadow all previous contents with the same dict key (without any message in the logs).
This also allows to drop in configuration examples from the documentation that modify multiple domains in a single file instead of splitting it to multiple files.

Add a new include tag "!include_dir_merge_recursive"
The tag will merge dicts recursively (if two or more files contain a dict with a mapping that has an identical key, both values will be merged). Dicts will be merged by applying the merge algorithm to the values of identical keys, lists will be merged by extending the existing lists. All other types lead to an exception. Recursion is stopped when hitting a value that is not a dict.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
!include_dir_merge_recursive configuration/
```

```yaml
# Example configuration/automation.yaml
automation: !include ../automations.yaml
```
```yaml
# Example configuration/devices/tp-link.yaml
tplink:
  discovery: true
  switch:
    - host: '10.13.37.1'
sensor:
  - platform: template
    sensors:
      plug_amps:
        friendly_name_template: "{{ states.switch.plug.name}} Current"
        value_template: '{{ states.switch.plug.attributes["current_a"] | float }}'
        unit_of_measurement: 'A'
recorder:
  exclude:
    entities:
      - switch.plug
```


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
This is kind of a request for comments at the moment, The PR works as is, but may need some improvements. For example I am considering adding references to all strings that are added to a dict from a merge. File-References are kind of wonky in check_config otherwise.

I will also do a separate pull request to add documentation if merging this PR is considered.

<!-- TODO: - Link to documentation pull request:  -->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

Phist0ne committed:
implement !include_dir_merge_recursive

The will merge dicts recursively (if two or more files contain a dict with a mapping that has an identical key, both values will be merged). dicts will be merged by applying the merge algorithm to the values of identical keys. lists will be merged by extending the existing lists. All other types lead to an exception. Recursion is stopped when hitting a value that is not a dict.

Phist0ne committed:
fix load_yaml converting empty lists to empty dicts

homeassistant commented:
Hi @Phist0ne, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31118).

Once you do that we will be able to review and accept this pull request.

Thanks!

Phist0ne committed:
add tests for !include_dir_merge_recursive tag

This tests a successful merge of both dicts and lists on first and second level as well as both error conditions "unsupported merge type" and "merging different types"

balloob commented:
I don't think that we should not add more YAML merge strategies. All of them are already complicated and hard to follow for YAML users.

@frenck what do you think?

frenck commented:
I do see the general use case for this, and on first look, it looks nice!

However, there are many cases where this can cause really unexpected behavior, mainly due to the merging of dicts. Some can be overridden without noticing it in the first place for example. And it becomes really hard to debug in terms of the order of things merged.

I think we should not support this use case for those reasons.

balloob commented:
Just thought of another reason not to merge this. A lot of config schemas allow a single configuration of a list of configurations. This can't be done with a simple merge command, you need to know if the config expects a dictionary or a list. It's what we do with packages merge logic, which is fragile at best.
